File: app/src/main/java/com/winnyking/wincalendar/tasker/htmlviewer/WebViewActivity.java
package com.winnyking.wincalendar.tasker.htmlviewer;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.winnyking.wincalendar.R;

public class WebViewActivity extends Activity {

    public static final String ACTION_FINISH = "com.winnyking.wincalendar.action.FINISH";
    public static final String EXTRA_VARIABLE_VALUES = "com.winnyking.wincalendar.extra.VARIABLE_VALUES";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_web_view);

        WebView webView = findViewById(R.id.webView);
        webView.getSettings().setJavaScriptEnabled(true);
        webView.setWebViewClient(new WebViewClient());

        Intent intent = getIntent();
        String htmlContent = intent.getStringExtra("html_content");
        String cssContent = intent.getStringExtra("css_content");
        String jsContent = intent.getStringExtra("js_content");
        String taskerVariables = intent.getStringExtra("tasker_variables");

        webView.addJavascriptInterface(new WebAppInterface(this, taskerVariables), "Android");

        String fullHtml = "<html><head><style>" + cssContent + "</style></head><body>" + htmlContent + "<script>" + jsContent + "</script></body></html>";
        webView.loadData(fullHtml, "text/html", "UTF-8");
    }

    public class WebAppInterface {
        Activity mActivity;
        String mTaskerVariables;

        WebAppInterface(Activity activity, String taskerVariables) {
            mActivity = activity;
            mTaskerVariables = taskerVariables;
        }

        @JavascriptInterface
        public void closeWebView(String values) {
            Intent intent = new Intent(ACTION_FINISH);
            intent.putExtra(EXTRA_VARIABLE_VALUES, values);
            sendBroadcast(intent);
            mActivity.finish();
        }
    }
}

File: app/src/main/java/com/winnyking/wincalendar/tasker/htmlviewer/WebViewActivity.kt
package com.winnyking.wincalendar.tasker.htmlviewer

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.webkit.JavascriptInterface
import android.webkit.WebViewClient
import com.winnyking.wincalendar.databinding.ActivityWebViewBinding

class WebViewActivity : Activity() {

    companion object {
        const val ACTION_FINISH = "com.winnyking.wincalendar.action.FINISH"
        const val EXTRA_VARIABLE_VALUES = "com.winnyking.wincalendar.extra.VARIABLE_VALUES"
    }

    private lateinit var binding: ActivityWebViewBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityWebViewBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.webView.settings.javaScriptEnabled = true
        binding.webView.webViewClient = WebViewClient()

        val html = intent.getStringExtra("html")

        binding.webView.addJavascriptInterface(WebAppInterface(this), "Android")

        binding.webView.loadDataWithBaseURL(null, html ?: "", "text/html", "UTF-8", null)
    }

    /**
     * This interface is exposed to the WebView and allows JavaScript code to interact with the Android app.
     * To return a value to Tasker, you can call the `closeWebView` method from your HTML's JavaScript.
     *
     * Example:
     * ```html
     * <button onclick="Android.closeWebView('valeur 1')">Close and Save</button>
     * ```
     * In this example, clicking the button will close the WebView and return the string 'valeur 1' to Tasker.
     * You can then use this value in your Tasker task.
     */
    inner class WebAppInterface(private val activity: Activity) {
        @JavascriptInterface
        fun closeWebView(values: String) {
            val intent = Intent(ACTION_FINISH).apply {
                putExtra(EXTRA_VARIABLE_VALUES, values)
            }
            sendBroadcast(intent)
            activity.finish()
        }
    }
}
